File: app/src/main/java/com/example/quoter/data/UserPreferencesRepository.kt
package com.example.quoter.data

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

// Define DataStore instance at the top level
private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "settings")

class UserPreferencesRepository(private val context: Context) {

    // Define Preference Keys
    private object PreferencesKeys {
        val LOCATION = stringPreferencesKey("pref_location")
        val SCENE = stringPreferencesKey("pref_scene")
        val STYLE = stringPreferencesKey("pref_style")
        val FONT_STYLE = stringPreferencesKey("pref_font_style") // New key for font style
        val FONT_SIZE = stringPreferencesKey("pref_font_size") // New key for font size
    }

    // Default values
    private val defaultLocation = "Mountains"
    private val defaultScene = "Natural"
    private val defaultStyle = "Photorealistic"

    // Flow to observe all preferences as a CustomizationState object
    val userPreferencesFlow: Flow<com.example.quoter.CustomizationState> = context.dataStore.data
        .map { preferences ->
            com.example.quoter.CustomizationState(
                location = preferences[PreferencesKeys.LOCATION] ?: defaultLocation,
                scene = preferences[PreferencesKeys.SCENE] ?: defaultScene,
                style = preferences[PreferencesKeys.STYLE] ?: defaultStyle,
                fontStyle = preferences[PreferencesKeys.FONT_STYLE] ?: "Bold",
                fontSize = preferences[PreferencesKeys.FONT_SIZE]?.toFloatOrNull() ?: 16f // Default to 16f if not set
            )
        }

    // --- Update Functions --- 

    suspend fun updateLocation(location: String) {
        context.dataStore.edit { preferences ->
            preferences[PreferencesKeys.LOCATION] = location
        }
    }

    suspend fun updateScene(scene: String) {
        context.dataStore.edit { preferences ->
            preferences[PreferencesKeys.SCENE] = scene
        }
    }

    suspend fun updateStyle(style: String) {
        context.dataStore.edit { preferences ->
            preferences[PreferencesKeys.STYLE] = style
        }
    }

    suspend fun updateFontStyle(fontStyle: String) {
        context.dataStore.edit { preferences ->
            preferences[PreferencesKeys.FONT_STYLE] = fontStyle
        }
    }

    suspend fun updateFontSize(fontSize: Float) {
        context.dataStore.edit { preferences ->
            preferences[PreferencesKeys.FONT_SIZE] = fontSize.toString()
        }
    }
}

File: swift_files/UserPreferencesRepository.swift
swift
import Foundation

struct CustomizationSettings {
    var location: String
    var scene: String
    var style: String
    var fontStyle: String
    var fontSize: Float
}

class UserPreferencesRepository {
    
    private let userDefaults: UserDefaults
    
    private struct PreferencesKeys {
        static let location = "pref_location"
        static let scene = "pref_scene"
        static let style = "pref_style"
        static let fontStyle = "pref_font_style"
        static let fontSize = "pref_font_size"
    }
    
    private let defaultLocation = "Mountains"
    private let defaultScene = "Natural"
    private let defaultStyle = "Photorealistic"
    
    init(userDefaults: UserDefaults = .standard) {
        self.userDefaults = userDefaults
    }
    
    var userPreferences: CustomizationSettings {
        get {
            return CustomizationSettings(
                location: userDefaults.string(forKey: PreferencesKeys.location) ?? defaultLocation,
                scene: userDefaults.string(forKey: PreferencesKeys.scene) ?? defaultScene,
                style: userDefaults.string(forKey: PreferencesKeys.style) ?? defaultStyle,
                fontStyle: userDefaults.string(forKey: PreferencesKeys.fontStyle) ?? "Bold",
                fontSize: userDefaults.float(forKey: PreferencesKeys.fontSize) == 0 ? 16.0 : userDefaults.float(forKey: PreferencesKeys.fontSize)
            )
        }
        set {
            userDefaults.set(newValue.location, forKey: PreferencesKeys.location)
            userDefaults.set(newValue.scene, forKey: PreferencesKeys.scene)
            userDefaults.set(newValue.style, forKey: PreferencesKeys.style)
            userDefaults.set(newValue.fontStyle, forKey: PreferencesKeys.fontStyle)
            userDefaults.set(newValue.fontSize, forKey: PreferencesKeys.fontSize)
        }
    }
    
    func updateLocation(location: String) {
        userDefaults.set(location, forKey: PreferencesKeys.location)
    }
    
    func updateScene(scene: String) {
        userDefaults.set(scene, forKey: PreferencesKeys.scene)
    }
    
    func updateStyle(style: String) {
        userDefaults.set(style, forKey: PreferencesKeys.style)
    }
    
    func updateFontStyle(fontStyle: String) {
        userDefaults.set(fontStyle, forKey: PreferencesKeys.fontStyle)
    }
    
    func updateFontSize(fontSize: Float) {
        userDefaults.set(fontSize, forKey: PreferencesKeys.fontSize)
    }
}